python/lsst/pipe/base/quantum_provenance_graph.py

```diff
@@ -1195,245 +1286,562 @@ def get_dataset_info(self, key: DatasetKey) -> DatasetInfo:
         """
         return self._xgraph.nodes[key]
 
-    def __add_new_graph(
+    def to_summary(
+        self, butler: Butler | None = None, do_store_logs: bool = True, n_cores: int = 1
+    ) -> Summary:
+        """Summarize the `QuantumProvenanceGraph`.
+
+        Parameters
+        ----------
+        butler : `lsst.daf.butler.Butler`, optional
+            Ignored; accepted for backwards compatibility.
+        do_store_logs : `bool`
+            Store the logs in the summary dictionary.
+        n_cores : `int`, optional
+
+        Returns
+        -------
+        result : `Summary`
+            A struct containing counts of quanta and datasets in each of
+            the overall states defined in `QuantumInfo` and `DatasetInfo`,
+            as well as diagnostic information and error messages for failed
+            quanta and strange edge cases, and a list of recovered quanta.
+        """
+        status_log = PeriodicLogger(_LOG)
+        if not self._finalized:
+            raise RuntimeError(
+                """resolve_duplicates must be called to finalize the
+                QuantumProvenanceGraph before making a summary."""
+            )
+        result = Summary()
+        futures: list[concurrent.futures.Future[None]] = []
+        _LOG.verbose("Summarizing %s tasks.", len(self._quanta.keys()))
+        with concurrent.futures.ThreadPoolExecutor(n_cores) as executor:
+            for m, (task_label, quanta) in enumerate(self._quanta.items()):
+                task_summary = TaskSummary()
+                task_summary.n_expected = len(quanta)
+                for n, quantum_key in enumerate(quanta):
+                    quantum_info = self.get_quantum_info(quantum_key)
+                    future = task_summary._add_quantum_info(
+                        quantum_info,
+                        log_getter=self._butler_get if do_store_logs else None,
+                        executor=executor,
+                    )
+                    if future is not None:
+                        futures.append(future)
+                    status_log.log(
+                        "Summarized %s of %s quanta of task %s of %s.",
+                        n + 1,
+                        len(quanta),
+                        m + 1,
+                        len(self._quanta.keys()),
+                    )
+                result.tasks[task_label] = task_summary
+            for n, future in enumerate(concurrent.futures.as_completed(futures)):
+                if (err := future.exception()) is not None:
```

Comment:
In theory you can collect all the exceptions from the futures and raise them in an `ExceptionGroup` when they've all completed.

Reply:
Thanks for the tip. In this case I don't think there are any error conditions we expect, and so it's better to fail fast because it means something is seriously wrong.

Reply:
I think that's right; I was thinking that an exception would make the context manager do `shutdown(wait=True, cancel_futures=True)` , but it seems it does `shutdown(wait=True, cancel_futures=False)` regardless of whether the context is exited with an exception or not.  So the result is that it runs to completion but shows the first error message.

That's okay in this context, as I think it's most likely they'd all have the same error here and 100k copies of the same error message is not a great experience.  But it's something we should think about if we add our own tooling for this; a context manager that does cancel on exceptions certainly seems more useful to me.